persist/fs/commitlog/reader.go

```diff
@@ -110,52 +157,132 @@ func (r *reader) Read() (
+		select {
+		case <-r.cancelCtx.Done():
+			for _, decoderBuf := range r.decoderBufs {
+				close(decoderBuf)
+			}
+			r.shutdownChan <- r.close()
 			return
+		default:
+			data, err := r.readChunk()
+			// TODO: Special case if err == io.EOF here?
+
+			// Distribute the decoding work in round-robin fashion so that when we
+			// read round-robin, we get the data back in the original order.
+			r.decoderBufs[index%r.numConc] <- decoderArg{
+				bytes: data,
+				err:   err,
+			}
+			index++
 		}
+	}
+}
 
-		id := r.bytesPool.Get(len(decoded.ID))
-		id.IncRef()
-		defer id.DecRef()
-		id.AppendAll(decoded.ID)
-
-		namespace := r.bytesPool.Get(len(decoded.Namespace))
-		namespace.IncRef()
-		defer namespace.DecRef()
-		namespace.AppendAll(decoded.Namespace)
-
-		r.metadataLookup[entry.Index] = Series{
-			UniqueIndex: entry.Index,
-			ID:          ts.BinaryID(id),
-			Namespace:   ts.BinaryID(namespace),
-			Shard:       uint32(decoded.Shard),
+func (r *reader) decoderLoop(inBuf <-chan decoderArg, outBuf chan *readResponse) {
+	decodingOpts := r.opts.FilesystemOptions().DecodingOptions()
+	decoder := msgpack.NewDecoder(decodingOpts)
+	metadataDecoder := msgpack.NewDecoder(decodingOpts)
+
+	for arg := range inBuf {
+		readResponse := &readResponse{}
+		if arg.err != nil {
+			readResponse.resultErr = arg.err
+			outBuf <- readResponse
+			continue
+		}
+		decoder.Reset(arg.bytes)
+		entry, err := decoder.DecodeLogEntry()
+		if err != nil {
+			readResponse.resultErr = err
+			outBuf <- readResponse
+			continue
 		}
-	}
 
-	metadata, ok := r.metadataLookup[entry.Index]
-	if !ok {
-		resultErr = errCommitLogReaderMissingLogMetadata
-		return
-	}
+		if len(entry.Metadata) != 0 {
+			metadataDecoder.Reset(entry.Metadata)
+			decoded, err := metadataDecoder.DecodeLogMetadata()
+			if err != nil {
+				readResponse.resultErr = err
+				outBuf <- readResponse
+				continue
+			}
+
+			id := r.bytesPool.Get(len(decoded.ID))
+			id.IncRef()
+			id.AppendAll(decoded.ID)
 
-	series = metadata
-	datapoint = ts.Datapoint{
-		Timestamp: time.Unix(0, entry.Timestamp),
-		Value:     entry.Value,
+			namespace := r.bytesPool.Get(len(decoded.Namespace))
+			namespace.IncRef()
+			namespace.AppendAll(decoded.Namespace)
+
+			r.metadataLookupLock.Lock()
+			r.metadataLookup[entry.Index] = Series{
+				UniqueIndex: entry.Index,
+				ID:          ts.BinaryID(id),
+				Namespace:   ts.BinaryID(namespace),
+				Shard:       uint32(decoded.Shard),
+			}
+			r.metadataLookupLock.Unlock()
+			namespace.DecRef()
+			id.DecRef()
+		}
+
+		numMetadataMiss := 0
+		var metadata Series
+		var ok bool
+		for true {
+			r.metadataLookupLock.Lock()
+			metadata, ok = r.metadataLookup[entry.Index]
+			r.metadataLookupLock.Unlock()
+
+			if ok {
+				break
+			}
+
+			// Race condition where the metadata is currently being processed by
+			// another decoderLoop goroutine, but that routine has not updated the
+			// metadataLookup map yet.
+			if !ok && numMetadataMiss < 3 {
```

Comment:
Hm this smells a little, metadata should be written once and only once in an entire commit log.  Did you run into this?  Also sleeping/timeout & backoffs does not provide strong deterministic behavior that we will need.